[AERIE-1910] - Scheduling EDSL Parameterless Activity Fix

pranav-super opened the pull request:
* **Tickets addressed:** AERIE-1910
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
This is a simple PR to address the concern in AERIE-1910 about parameterless activities allowing bogus parameters. So, if there was an activity, say `EmptyActivity`, admitting no parameters (meaning the correct instantiation would be `EmptyActivity({})`), the EDSL could allow `EmptyActivity({bogus: "param"})` to be passed. To avoid that, the fix this PR takes per @mattdailis's recommendation, is just to not accept any parameters. That is, instead of instantiating with an empty dictionary (`EmptyActivity({})`), now, no dictionary is taken (`EmptyActivity()`).

## Verification
Tests were created! A new test was created in the compilation tester, and the integration tester, and it was tested with a deployment of AERIE to confirm that the code editor could suggest this change as well.

## Documentation
A quick update to the **ActivityTemplate** section was done.

## Future work
None, maybe updating the constraints eDSL to also handle this?

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/TypescriptCodeGenerationService.java:
```diff
@@ -86,11 +86,35 @@ private static String generateActivityTemplateConstructors(final Iterable<Activi
-      result.add(indent("}): %s {".formatted(activityTypeCode.activityTypeName())));
-      result.add(indent(indent("return { activityType: ActivityType.%s, args };".formatted(activityTypeCode.activityTypeName()))));
-      result.add(indent("},"));
+      if(activityTypeCode.parameterTypes().size() == 0) {
```
Prefer `.isEmpty()` to `.size() == 0`
```suggestion
      if(activityTypeCode.parameterTypes().isEmpty()) {
```

pranav-super replied:
Fixed!

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/TypescriptCodeGenerationService.java:
```diff
@@ -86,11 +86,35 @@ private static String generateActivityTemplateConstructors(final Iterable<Activi
-      result.add(indent("},"));
+      if(activityTypeCode.parameterTypes().size() == 0) {
+        /*
+        Relax: function RelaxConstructor(): Relax {
```
This comment looks like it was helpful when writing this code, but I'm worried it will mostly be confusing in the future. For one, it's already out of date (the returned dictionary is missing the `args` key) and it will only get more out of date as we continue to make changes to this code. I suggest removing these comments, but making sure that the `TypescriptCodeGenerationTest` has examples of activities with and without parameters.

pranav-super replied:
I removed it!

mattdailis commented on scheduler-server/src/test/java/gov/nasa/jpl/aerie/scheduler/server/services/SchedulingIntegrationTests.java:
```diff
@@ -697,6 +698,31 @@ export default (): Goal => {
   }
 
+  @Test
+  void testEmptyActivity() {
```
What is this test testing that isn't already covered by the new unit tests? I think we can remove this one.

pranav-super replied:
I guess just that none of them test an empty activity as BananaNation doesn't have any. I thought it would be good to prove that this update doesn't break anything in the integration tests but I don't disagree that it's extraneous. I'll go ahead and remove it!

pranav-super replied:
Removed!

mattdailis commented on examples/banananation/src/main/java/gov/nasa/jpl/aerie/banananation/activities/RelaxActivity.java:
```diff
@@ -0,0 +1,27 @@
+
+  @EffectModel
+  public void run(final Mission mission) {
+    //delay(Duration.of(RELAX_DURATION, Duration.HOURS));
```
I think we can delete this line, as well as the `RELAX_DURATION` field.

pranav-super replied:
Sounds good. I wasn't sure about how to use `delay()` for this model so I included it as a placeholder but I can get rid of it. It doesn't seem like any of the activities in BananaNation have a duration, which I thought might be useful to test but I'm unsure how to do that.

pranav-super replied:
Removed!

mattdailis reviewed (changes requested):
Generally looks solid to me. Just a few minor code cleanup requests.

skovati commented on examples/banananation/src/main/java/gov/nasa/jpl/aerie/banananation/activities/RelaxActivity.java:
```diff
@@ -0,0 +1,27 @@
+import gov.nasa.jpl.aerie.banananation.Mission;
+import gov.nasa.jpl.aerie.merlin.framework.annotations.ActivityType;
+import gov.nasa.jpl.aerie.merlin.framework.annotations.ActivityType.EffectModel;
+import gov.nasa.jpl.aerie.merlin.framework.annotations.Export.Parameter;
+import gov.nasa.jpl.aerie.merlin.framework.annotations.Export.Validation;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+
+import java.util.List;
```
```suggestion
```
Looks like these imports are unused.

pranav-super replied:
Facts

skovati reviewed (changes requested):
Looks pretty good otherwise!

pcrosemurgy commented on examples/banananation/src/main/java/gov/nasa/jpl/aerie/banananation/activities/RelaxActivity.java:
```diff
@@ -0,0 +1,19 @@
+ * Relax? I needed something with no parameters this was the best one I could come up with
+ *
+ * @subsystem fruit
+ * @contact Jane Doe
```
Pedantic since these `@` javadoc lines are harmless but I don't see a compelling reason to check them into VCS

pcrosemurgy replied:
Also, this could be called `EffectlessActivity` if you're looking for another name

skovati replied:
I agree the name could be more descriptive, preferably something banana / monke themed...

pranav-super replied:
So should I just remove the @'s? It seems like all the other activities have them so I added them for the sake of uniformity. I can rename the activity to something better too.

pcrosemurgy replied:
Ahh true! Didn't realize the others do. Nice, in that case I think what you have is the right thing to do. :)

Twisol replied:
FYI, the *reason* those nonstandard Javadoc tags are present is because we originally (and, I think, still?) wanted to extract that information out at annotation processing time, so it could be flowed out to planners via the UI.

That hasn't happened yet, so these tags don't do anything. Their presence is a good reminder that we still have yet to implement that functionality... but only, of course, if we knew it was a thing to begin with 😁

pranav-super replied:
How's BananaNapActivity

pranav-super replied:
Fixed!

pcrosemurgy commented on examples/banananation/src/main/java/gov/nasa/jpl/aerie/banananation/activities/RelaxActivity.java:
```diff
@@ -0,0 +1,19 @@
+public final class RelaxActivity {
+  @EffectModel
+  public void run(final Mission mission) {
+    return;
```
Can safely remove this `return`

pranav-super replied:
Removed!

Commit:
add tests

Commit:
update code generation

Commit:
add empty banananation activity

While foomissionmodel has BarActivity, BananaNation has none. This was a problem when writing integration tests (which ultimately got deleted as they were redundant), but I figured it was worth including for future testing regardless.
